src/Products/Viewer/Controllers/ViewerApiController.cs

```diff
@@ -674,97 +548,140 @@ private static LoadDocumentEntity GetDocumentPages(PostedDataEntity postedData,
             // get/set parameters
             string documentGuid = postedData.guid;
             string password = string.IsNullOrEmpty(postedData.password) ? null : postedData.password;
-            if (!File.Exists(documentGuid)) 
-            { 
-                throw new GroupDocsViewerException("File not found."); 
+
+            var fileFolderName = Path.GetFileName(documentGuid).Replace(".", "_");
+            string fileCacheSubFolder = Path.Combine(cachePath, fileFolderName);
+            string htmlResourcePrefix = "/viewer/resources/";
+
+            if (!File.Exists(documentGuid))
+            {
+                throw new GroupDocsViewerException("File not found.");
             }
 
-            string cachePath;
-            ViewerSettings settings = GetViewerSettings(documentGuid, out cachePath);
+            IViewerCache cache = new FileViewerCache(cachePath, fileCacheSubFolder);
 
-            using (GroupDocs.Viewer.Viewer viewer = new GroupDocs.Viewer.Viewer(documentGuid, GetLoadOptions(password), settings))
+            LoadDocumentEntity loadDocumentEntity = new LoadDocumentEntity();
+            if (globalConfiguration.Viewer.GetIsHtmlMode())
             {
-                if (loadAllPages)
+                using (HtmlViewer htmlViewer = new HtmlViewer(documentGuid, htmlResourcePrefix, cache, GetLoadOptions(postedData.password)))
                 {
-                    GenerateViewerCache(viewer);
+                    loadDocumentEntity = GetLoadDocumentEntity(loadAllPages, documentGuid, password, fileCacheSubFolder, htmlViewer);
                 }
-
-                dynamic viewInfo;
-                LoadDocumentEntity loadDocumentEntity = new LoadDocumentEntity();
-
-                if (!Directory.Exists(cachePath))
+            }
+            else
+            {
+                using (PngViewer pngViewer = new PngViewer(documentGuid, htmlResourcePrefix, cache, GetLoadOptions(postedData.password)))
                 {
-                    Directory.CreateDirectory(cachePath);
+                    loadDocumentEntity = GetLoadDocumentEntity(loadAllPages, documentGuid, password, fileCacheSubFolder, pngViewer);
                 }
+            }
 
-                var pagesInfoPath = Path.Combine(cachePath, "PagesInfo.xml");
-                viewInfo = viewer.GetViewInfo(ViewInfoOptions.ForHtmlView());
+            return loadDocumentEntity;
+        }
 
-                if (!File.Exists(pagesInfoPath))
-                {
-                    var xdoc = new XDocument(new XElement("Pages"));
+        private static LoadDocumentEntity GetLoadDocumentEntity(bool loadAllPages, string documentGuid, string password, string fileCacheSubFolder, ICustomViewer customViewer)
```

Comment:
Codacy found an issue: [Remove this unused method parameter 'password'.](https://app.codacy.com/manual/GroupDocs/GroupDocs.Viewer-for-.NET-MVC/pullRequest?prid=5933706)